src/main/java/spoon/reflect/visitor/OperatorHelper.java

```diff
@@ -219,4 +244,277 @@ public static OperatorAssociativity getOperatorAssociativity(UnaryOperatorKind o
+					}
+
+					return Optional.empty();
+				});
+			}
+			case LT:
+			case LE:
+			case GT:
+			case GE:
+			case MUL:
+			case DIV:
+			case MOD:
+			case MINUS:
+				// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.20
+				return binaryNumericPromotion(left, right);
+			case PLUS:
+				return binaryNumericPromotion(left, right).or(() -> {
+					// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.18.1
+					//
+					// If the type of either operand of a + operator is String, then the operation is
+					// string concatenation.
+					CtTypeReference<?> stringType = typeFactory.STRING;
+					if (left.getType().equals(stringType) || right.getType().equals(stringType)) {
+						return Optional.of(stringType);
+					}
+
+					return Optional.empty();
+				});
+			case BITAND:
+			case BITXOR:
+			case BITOR: {
+				// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.22
+				CtTypeReference<?> leftType = left.getType().unbox();
+				CtTypeReference<?> rightType = right.getType().unbox();
+
+				Set<CtTypeReference<?>> floatingPointNumbers = Set.of(
+					typeFactory.FLOAT_PRIMITIVE,
+					typeFactory.DOUBLE_PRIMITIVE
+				);
+				if (floatingPointNumbers.contains(leftType) || floatingPointNumbers.contains(rightType)) {
+					return Optional.empty();
+				}
+
+				if (leftType.equals(rightType) && leftType.equals(typeFactory.BOOLEAN_PRIMITIVE)) {
+					return Optional.of(leftType);
+				}
+
+				return binaryNumericPromotion(left, right);
+			}
+			default:
+				throw new IllegalArgumentException("Unknown operator: " + operator);
+		}
+	}
+
+	/**
+	 * Get the promoted type of an unary operator, as defined by the Java Language Specification.
+	 *
+	 * @param operator the operator
+	 * @param operand the operand
+	 * @return the promoted type or {@link Optional#empty()} if promotion does not apply or the operation is invalid
+	 */
+	public static Optional<CtTypeReference<?>> getPromotedType(
+		UnaryOperatorKind operator,
+		CtExpression<?> operand
+	) {
+		TypeFactory typeFactory = operand.getFactory().Type();
+		CtTypeReference<?> operandType = operand.getType();
+		switch (operator) {
+			case COMPL:
+				if (operandType.unbox().isPrimitive()
+					&& (WHOLE_NUMBERS.contains(operandType.unbox().getActualClass())
+					|| operandType.unbox().equals(typeFactory.CHARACTER_PRIMITIVE))) {
+					return unaryNumericPromotion(operand);
+				}
+
+				return Optional.empty();
+			case POS:
+			case NEG:
+				// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.15.3
+				return unaryNumericPromotion(operand);
+			case NOT:
+				if (operand.getType().unbox().equals(typeFactory.BOOLEAN_PRIMITIVE)) {
+					return Optional.of(typeFactory.BOOLEAN_PRIMITIVE);
+				}
+
+				return Optional.empty();
+			case PREINC:
+			case PREDEC:
+			case POSTINC:
+			case POSTDEC:
+				// Not yet implemented, because it is not necessary for the current use case.
```

Comment:
This should be either addressed or documented in an issue so we can keep track of it.